File: SantanderLago/pagarapi/validators.py
from decimal import Decimal, DecimalException
from .models import Account
from .utils import decompose_cbu


def validate_cbu_get_account(cbu, must_exist=False) -> tuple[(Account | None), (str | None)]:
    """Validates whether a CBU is valid and from a local account that exists. Returns the account or an error message"""
    entity_number, branch_number, account_number, is_ok = decompose_cbu(cbu)
    if not is_ok:
        return None, "CBU Checksum doesn't match"

    try:
        return Account.objects.get_by_cbu(cbu), None
    except Account.DoesNotExist:
        return None, f"CBU {cbu} does not exist locally" if must_exist else None


def validate_transaction_amount(amount) -> tuple[(Decimal | None), (str | None)]:
    try:
        amount = Decimal(amount)
    except DecimalException:
        return None, "Not a valid amount"
    if amount <= 0:
        return None, "Must be >= 0"
    if not amount == amount.quantize(Decimal('1.00')):
        return None, 'Must have at most two decimal digits'

    return amount, None

File: Galicio/validators/validators.js
const cbuUtils = require("../cbuUtils")
const dotenv = require('dotenv');
const Account = require("../models/Account");
const BigDecimal = require("../bigdecimal");

async function validateCbuAccount(cbu, isSource, amountd) {
    const {entityNumber, branchNumber, accountNumber, isOk} = cbuUtils.decompose(cbu);

    if (!isOk) {
        throw new Error("CBU Checksum doesn't match")
    }

    const account = await Account.find().getByCbu(cbu).exec();

    if (!account) {
        throw new Error((isSource ? "Source" : "Destination") + " account doesn't exist");
    }

    const isCentral = branchNumber == 0 && accountNumber == 0;
    if (isSource) {
        if (!isCentral && amountd.isGreaterThan(account.balance)) {
            throw new Error("Insufficient Balance");
        }
    } else {
        if (!account.active) {
            throw new Error("Cannot transfer money to a deleted account");
        }
    }

    return account;
}

function validateTransactionAmount(amount) {
    amount = amount.toString();

    if (!amount.match(/^[0-9]+(.[0-9]+)?$/)) {
        throw new Error('Not a valid amount');
    }

    const decimal = new BigDecimal(amount);

    if (decimal.isLesserOrEqualThan(0)) {
        throw new Error('Amount must by greater than 0');
    }

    const amountString = amount.toString();
    const decimalDigits = (amountString.split('.')[1] || '').length;

    if (decimalDigits > 2) {
        throw new Error('Amount must have at most two decimal digits');
    }

    return decimal;
}

module.exports.validateTransactionAmount = validateTransactionAmount;
module.exports.validateCbuAccount = validateCbuAccount;
